### ISSUE TITLE
FormSection and nested Field component bug

### ISSUE THREAD

**Issue by Kestutis** [erikras/redux-form#2852]
### Are you submitting a **bug report** or a **feature request**?
Bug with FormSection and nested Field

### What is the current behavior?
When using FormSection with nested Field components, input names get prefix with every nested reuse. 
```javascript 

const fieldComponent = ({input}) => {
	console.log(input.name); // as expected form1.foo
        return <Field {...input} component={nextFieldComponent} />
}

const nextFieldComponent = ({input}) => {
	console.log(input.name); // becomes form1.foo.foo
	return <input {...input} />
}

const Form = () => (
  <div>
    <FormSection name='form1'>
      <Field name='foo' component={fieldComponent} />
    </FormSection>    
  </div>
)
```
fiddle with alert on this: https://jsfiddle.net/86v1yuuz/2/

### What is the expected behavior?
It should not append FormSection name on nested Field components

### Sandbox Link
https://jsfiddle.net/86v1yuuz/2/

### What's your environment?
Redux v. 6.6.3, React 15.5.3

**Comment by cyberpluto** [comment]
We are seeing the same issue with `<FormSection>` nested inside `<FormSection>`

**Comment by wmertens** [comment]
This causes problems when you have e.g.

```js
const Section = ({input: {name, value}}) => (
  <FormSection name={name}>
      <Field Component={Toggle} name="enableSub"/>
      {value.enableSub && <Field Component={Section} name="sub"/>}
  </FormSection>
)
```

Nesting like this will work ok for the first Section (say it has the name "root"), but every subSection will add the full path to the `sectionPrefix`, because `Field` will pass the full `"root.sub"` name to Section instead of the given `"sub"`.

It seems to me that either:
* Field components should always get their `name` relative to `sectionPrefix`
* or FormSection should accept an `absolute` parameter so it can replace `sectionPrefix` entirely, e.g. `<FormSection absolute name={name}>` in the example

The former is more intuitive IMHO but is an API change that could cause issues. The latter is a simpler solution, and can be used in other situations as well.

**Comment by wmertens** [comment]
Another option would be to have to add a `.` in front of the name to make the `FormSection` relative.

```js
<FormSection name="root">
  <FormSection name=".sub"/> // sectionPrefix becomes root.sub
  <FormSection name="otherRoot"/> // sectionPrefix becomes otherRoot
</FormSection>
```

**Comment by erikras** [comment]
It would never have occurred to me to use `<Field>` inside a `component` rendering prop giving to `<Field>`. Feeding the `props.input` generated by `<Field>` back into itself as props makes no sense to me. I'm struggling to understand why someone would make this design decision.

**Comment by wmertens** [comment]
This is great for nesting components, thus creating recursive forms where
you don't know beforehand if the component will be a simple or complex
field.

**Comment by wmertens** [comment]
A very concrete example: editing a template page where you can change
sections of the page, and sections either showing their current value or
being edited as well and showing editing fields, including nesting sections.

Is there a better way to do this besides nesting Fields?

**Comment by erikras** [comment]
> Is there a better way to do this besides nesting Fields?

This sounds like the use case that `<FormSection>` was designed for. The canonical example is using a single `<AddressFields/>` component for separate "sections" of your data.

```jsx
const AddressFields = () => (
  <Field name="street" component="input"/>
  <Field name="city" component="input"/>
  <Field name="postalCode" component="input"/>
)

<FormSection name="billing"><AddressFields/></FormSection>
<FormSection name="shipping"><AddressFields/></FormSection>
```
...so you end up with...
```jsx
{
  billing: {
    street: '123 Main St.',
    city: 'Los Angeles',
    postalCode: '90210'
  },
  shipping: {
    street: '42 Fifth Ave.',
    city: 'New York',
    postalCode: '10011'
  }
}
```

**Comment by wmertens** [comment]
Yes, each SectionField is nesting its own FormSection when it's in edit
mode. Turtles all the way down :)

**Comment by erikras** [comment]
IMHO, `Field`s should never be nested. A _field_ is the input component. Have a complex input component? Fine, but the input component should not need a `<Field>` component inside of it. Perhaps this is not obvious and should generate a warning?

`FormSection` and `FieldArray` provide ways of structuring object and array data respectively.

**Comment by wmertens** [comment]
Hmmm now that you explain it like that, in our case we can do that by making per-section SubForms instead of per-section complex Fields. Not that big of a change really, and then nesting happens with FormSection wrapping inside the SubForm. Indeed, that would be better, and I wonder if something similar could be done by @Kestutis ?

**Comment by Kestutis** [comment]
in my case, if I remember this right, it was like we have lots of default styled components(almost like UI styleguide) and sometimes we need to get into more complex components out of these defaults.
so If you have `defaultSelectComponentField` with all of the default styling and behavior, and react-select implementation, but then you need something with different labeling maybe, or some extra wrapping or extra error control and you want to reuse this component instead of copying it.

**Comment by wmertens** [comment]
@erikras I tried replacing Fields holding Fields with forms holding Fields, and it almost works, except when the form needs the value of a field. In our case, the value we need is the type of the current section so we can know which form component to load next.

The `formValueSelector` is not useful because it doesn't read the current `sectionPrefix`.

As a workaround, we (ab)use `format`  to get the current value as in callback and assign that to state:

```js
<Field name="type" component={() => null} format={type => setTimeout(() => this.setState({type}), 0)}/>
```

This is pretty hacky, it doesn't know the type until the first render, but it works at least. The `setTimeout` is needed to prevent state setting during render.

It would be nicer if we could have a `@formValue('type')` decorator, which would give `type` as a prop, automatically using the correct formname and sectionPrefix. Thoughts?

**Comment by erikras** [comment]
> As a workaround, we (ab)use format to get the current value as in callback and assign that to state:

Great, now there's vomit all over my keyboard! 🤢 

> It would be nicer if we could have a `@formValue('type')` decorator, which would give `type` as a prop, automatically using the correct formname and sectionPrefix. Thoughts?

Okay, so if you are in a component that has a collection of fields (e.g. `street`, `city`, etc.) you want a decorator HOC (I would call it `formPath` or `formPrefix`) that will give you the string that will be prefixed onto the name of each of your fields? e.g. `'billing'`, from my example above. The singular param to the decorator would be the name of the prop to inject it into (although perhaps it could default to `formPath`).

Am I understanding you correctly?

**Comment by wmertens** [comment]
> Okay, so if you are in a component that has a collection of fields (e.g. street, city, etc.) you want a decorator HOC (I would call it formPath or formPrefix) that will give you the string that will be prefixed onto the name of each of your fields? e.g. 'billing', from my example above. The singular param to the decorator would be the name of the prop to inject it into (although perhaps it could default to formPath).

No, we definitely need the value, not the path. We need a way to get at the current value of `${sectionPrefix}.someVar` without being a Field. This is the only way to do recursive forms that rely on the form value to show the form but are not nesting Fields in Fields.

**Comment by erikras** [comment]
I was thinking that _with the path_, you could then use the existing `connect()`/`formValueSelector()` system to fetch the value. Too bad `connect()` doesn't have a `mapContextToProps` option. 😆 

How about:
```jsx
@formValues({
  myType: 'typeField'
})
```
where `myType` is the prop the value would be injected as, and `typeField` is the name of the field _relative to the current form path_ to fetch the value for?

I really rather like the simplicity of providing _selectors_ rather than decorators, but for this we need context access.

**Comment by wmertens** [comment]
How about both `@formValues(name, name, ...)` and `@formValues({propName: fieldName, ...})` (like you propose)? Doesn't cost anything extra to support both… If not, then what you propose.

**Comment by wmertens** [comment]
BTW @Kestutis I think you can solve your problem as follows:

```js
const FieldComponent = ({name}) => {
	console.log(name); // as expected foo
        return <Field name={name} component={NextFieldComponent} />
}

const NextFieldComponent = ({input}) => {
	console.log(input.name); // becomes form1.foo
	return <input {...input} />
}

const Form = () => (
  <div>
    <FormSection name='form1'>
      <FieldComponent name='foo'/>
    </FormSection>    
  </div>
)
```

Try to make your forms so that no matter what the structure is, the top component is the one that has `@reduxForm`, all intermediate components use only `<FormSection>` if they need to work on subtrees, and the `<Field ...>` components are at the bottom of your structure.

**Comment by erikras** [comment]
> the `<Field ...>` components are at the bottom of your structure.

Yes! `<Field>`'s are leaves! 🌳

> Also, React components need to be uppercase, otherwise JSX won't recognize them as components.

They don't have to be if they are values passed to a `component` prop. But the argument that they should be, to denote that they are _components_, could reasonably be made. I'd have a lot of docs to change to adopt that viewpoint. 😄 

Closing this. `formValues()` discussion may/should continue on #2996.

**Comment by erikras** [comment]
Fix published in [`v6.8.0`](https://github.com/erikras/redux-form/releases/tag/v6.8.0).
